Add guide to raise a patch in Lighthouse. (#121)
submariner-io/submariner-website: src/content/contributing/lighthouse/_index.en.md

> ---

Reviewer: @aswinsuryan please use make server locally in your laptop to verify the rendering, it's not working in a few places (see comments bellow).

Author: Oh I missed that step for this one, thanks for checking it. Will verify it.

Author: Fixed the issues.

> title: "Contributing to Lighthouse"

Author: Added line wraps.

```
git clone https://github.com/<your-github-username>/lighthouse.git
cd lighthouse
git checkout -b <BRANCH-NAME>
```

####Contrubuting to Lighthouse Agent

The Lighthouse agent code is present at pkg folder inside the Lighthouse house directory. There is a package for Kubernetes controller which listens for service creation and it also has helpers for creating MultiClusterService CRD. Once you have made the code changes you can refer [building](https://submariner-io.github.io/contributing/building_testing/#submariner-iolighthouse) and [testing] (https://submariner-io.github.io/contributing/building_testing/#common-build-and-testing-targets) to check if the changes have no regression.

Once you run the make e2e you will have a 3 virtual cluster deployment, you could use it to verify if the changes are working as expected manually. Lighthouse agent will be running the submariner-operator namespace in the data clusters. You could use kubectl commands to see the logs and the pods status.

####Contrubuting to Lighthouse Plugin

> The plugin/lighthouse/ folder has the lighthouse plugin code. This uses the structure of as required by a standard code DNS plugin. The plugin answers the DNS requests by using the info in multicluster service. After making changes you can [build](https://submariner-io.github.io/contributing/building_testing/#submariner-iolighthouse) and [test](https://submariner-io.github.io/contributing/building_testing/#common-build-and-testing-targets) the code.

Reviewer: ```suggestion
The plugin/lighthouse/ folder has the lighthouse plugin code. This uses the structure of as required by a standard code DNS plugin. The plugin answers the DNS requests by using the info in multicluster service. After making changes you can [build](../building_testing/#submariner-iolighthouse) and [test](.../building_testing/#common-build-and-testing-targets) the code.
```

---
title: "Contributing to Lighthouse"
date: 2020-04-20T17:12:52+0300
weight: 10
---

The [Lighthouse Project](https://github.com/submariner-io/lighthouse) has an agent and a coreDNS plugin.
You can find more about the architecture at [Lighthouse Architecture](../../architecture/components/lighthouse/).

> This guide helps in making your contribution easier.

Author: Done.

### Fork and get the codebase

You need to fork the [Lighthouse repo](https://github.com/submariner-io/lighthouse/fork) after logging into your GitHub account.
Then you can clone your forked repo locally and switch to the lighthouse folder. You need to create a branch in order to raise a PR

```
git clone https://github.com/<your-github-username>/lighthouse.git
cd lighthouse
git checkout -b <BRANCH-NAME>
```

#### Contributing to Lighthouse Agent

> The Lighthouse agent code is present at pkg folder inside the Lighthouse house directory. There is a package for Kubernetes controller
> which listens for service creation and creates MultiClusterService CRs. Once you have made the code changes you can refer [building](../building_testing/#submariner-iolighthouse)
> and [testing](../building_testing/#common-build-and-testing-targets) it to check if the changes have no regression.

Reviewer 2: ```suggestion
The Lighthouse Agent code is located under the `pkg/agent` directory. The `Controller` is the main component that performs bi-directional syncing of Service and MultiClusterService resources. After making code changes, please refer to [building](../building_testing/#submariner-iolighthouse)
and [testing](../building_testing/#common-build-and-testing-targets) to build and verify your changes.
```

> Once you run the make e2e you will have a 3 virtual cluster deployment, you could use it to verify if the changes are working as expected manually.
> Lighthouse agent will be running the submariner-operator namespace in the data clusters. You could use kubectl commands
> to see the logs and the pods status.

Reviewer 2: ```suggestion
Running `make e2e` will create a 3-cluster deployment which you can use it manually verify changes if necessary.
The Lighthouse Agent will be deployed in the `submariner-operator` namespace in each cluster. You can use `kubectl` commands to check the pod's logs and status.
```

#### Contributing to Lighthouse Plugin

> The plugin/lighthouse/ folder has the lighthouse plugin code. This uses the structure of as required by a standard code DNS plugin.
> The plugin answers the DNS requests by using the info in multicluster service. After making changes you can [build](../building_testing/#submariner-iolighthouse)
> and [test](.../building_testing/#common-build-and-testing-targets) the code.

Reviewer 2: ```suggestion
The Lighthouse plugin code is located under the `plugin/lighthouse` directory and uses the structure as required by a standard code CoreDNS plugin.
The plugin answers the DNS requests by using the info in multicluster service. After making code changes, please refer to [building](../building_testing/#submariner-iolighthouse)
and [testing](../building_testing/#common-build-and-testing-targets) to build and verify your changes.
```

> Similar to the agent, you can use the deployment created using make e2e. The lighthouse plugin will be running as a part of the coreDNS in the kube-system 
> namespace and the config map will be changed to use it. The usual kubectl commands can be used here as well.
> You can add the debug entry in the coreDNS config map if you require more logging.

Reviewer 2: ```suggestion
Similar to the agent, you can use the deployment created using `make e2e` to manually test and verify as well. The Lighthouse plugin will be running as a part of the CoreDNS in the `kube-system`  
namespace and the CoreDNS config map will be changed to use it. The usual `kubectl` commands can be used here as well.
You can add the debug entry in the CoreDNS config map if you require more logging.
```

> ### Raising a Pull Request

Reviewer 2: ```suggestion
### Creating a Pull Request
```

> You can commit your changes using git commit and push the changes to your fork branch. Then you will see an option to raise a PR in
> the GitHub UI in your browser when you go to your forked repo.

Reviewer 2: ```suggestion
You can commit your changes using `git commit` and push the changes to your fork branch. Then you will see an option to create a PR in
the GitHub UI in your browser when you go to your forked repository.
```

```
git commit -s #Give a commit message
git push -f origin HEAD:<Your-Branch-Name>
```

> Raising a PR will trigger CI and will verify the changes.

Reviewer 2: ```suggestion
Creating a PR will trigger CI to verify the changes.
```

#### Contributing to Lighthouse Agent

The Lighthouse Agent code is located under the `pkg/agent` directory. The `Controller` is the main component that performs bi-directional syncing of Service and MultiClusterService resources. After making code changes, please refer to [building](../building_testing/#submariner-iolighthouse)
and [testing](../building_testing/#common-build-and-testing-targets) to build and verify your changes.

> Running `make e2e` will create a 3-cluster deployment which you can use it manually verify changes if necessary.

Reviewer 2: ```suggestion
Running `make e2e` will create a 3-cluster deployment which you can use to manually verify changes if necessary.
```

The Lighthouse Agent will be deployed in the `submariner-operator` namespace in each cluster. You can use `kubectl` commands to check the pod's logs and status.

#### Contributing to Lighthouse Plugin

> The Lighthouse plugin code is located under the `plugin/lighthouse` directory and uses the structure as required by a standard code CoreDNS plugin.

Reviewer 2: ```suggestion
The Lighthouse plugin code is located under the `plugin/lighthouse` directory and uses the structure as required by a standard CoreDNS plugin.
```

> The plugin answers the DNS requests by using the info in multicluster service. After making code changes, please refer to [building](../building_testing/#submariner-iolighthouse)

Reviewer 2: ```suggestion
The plugin answers DNS requests by using the MultiClusterService resources populated by the Lighthouse Agent. After making code changes, please refer to [building](../building_testing/#submariner-iolighthouse)
```
